network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,356 @@
+            let now = now.to_le_bytes().to_vec();
+            Some(now)
+        } else {
+            None
+        };
+
+        // create first handshake message  (-> e, es, s, ss)
+        let mut rng = rand::rngs::OsRng;
+        let mut first_message = [0u8; noise::handshake_init_msg_len(PAYLOAD_SIZE)];
+        let initiator_state = self
+            .0
+            .initiate_connection(
+                &mut rng,
+                &[],
+                remote_public_key,
+                payload.as_ref().map(|x| &x[..]),
+                &mut first_message,
+            )
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("noise: wrong input passed {}", e),
+                )
+            })?;
+
+        // write the first handshake message
+        socket.write_all(&first_message).await?;
+
+        // flush
+        socket.flush().await?;
+
+        // receive the server's response (<- e, ee, se)
+        let mut server_response = [0u8; noise::handshake_resp_msg_len(0)];
+        socket.read_exact(&mut server_response).await?;
+
+        // parse the server's response
+        // TODO: security logging here? (mimoo)
+        let (_, session) = self
+            .0
+            .finalize_connection(initiator_state, &server_response)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // finalize the connection
+        Ok(NoiseStream::new(socket, session))
+    }
+
+    pub async fn accept<TSocket>(
+        &self,
+        mut socket: TSocket,
+        anti_replay_timestamps: Option<Arc<RwLock<AntiReplayTimestamps>>>,
+        trusted_peers: Option<&Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>>,
+    ) -> io::Result<NoiseStream<TSocket>>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // receive the initiation message
+        let mut client_init_message = [0u8; noise::handshake_init_msg_len(PAYLOAD_SIZE)];
+        socket.read_exact(&mut client_init_message).await?;
+
+        // parse it
+        let (their_public_key, handshake_state, payload) = self
+            .0
+            .parse_client_init_message(&[], &client_init_message)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // make sure the public key is a validator before continuing (if we're in the validator network)
+        if let Some(trusted_peers) = trusted_peers {
+            let found = trusted_peers
+                .read()
+                .unwrap()
+                .iter()
+                .any(|(_peer_id, public_keys)| public_keys.identity_public_key == their_public_key);
```

Comment:
That's a `find` you're looking for here.

Reply:
`any() =  find() + some()`